### PR TITLE
Document and implement nn::gfx

### PR DESCRIPTION
**Pull request by ThePixelGamer**
At the moment only the fields are semi documented, I want the functions to be implemented before merging

- - -
This change is [Reviewable](https://reviewable.io/reviews/open-ead/nnheaders/12)

### CODE REVIEW

**Review by leoetlino — COMMENTED** [review]
Function definitions should probably go into a .cpp file.

Also, formatting is off.

**Comment by ThePixelGamer** [comment]
Blocked by #14

File: include/nn/util.h [hunk 1]
```diff
@@ -58,3 +58,18 @@ struct BitFlagSet {};
 
 void ReferSymbol(const void*);
 }  // namespace nn
+
+// todo: move with more info
+#define NN_NO_COPY(CLASS)                                                                          \
+private:                                                                                           \
```
Comment (leoetlino) [inline include/nn/util.h:64]: now that we're using `= delete;` I think `private:` isn't really necessary anymore (and removing it lets us get rid of a gotcha)

Comment (ThePixelGamer) [reply include/nn/util.h:64]: It's necessary, the types have those functions as private

**Comment by ThePixelGamer** [comment]
All the functions I want to match are now matching, it's now fully ready for a code review 👍

File: include/nn/util/util_TypedStorage.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <nn/types.h>
+#include <type_traits>
+
+namespace nn::util {
+
+template <class T, size_t Size, size_t Alignment>
+struct TypedStorage {
+    typename std::aligned_storage<Size, Alignment>::type storage;
```
Comment (leoetlino) [inline include/nn/util/util_TypedStorage.h:10]: As previously discussed, std::aligned_storage is deprecated in modern standards so we should avoid using it (even if that's what the original nn code does). `std::aligned_storage` was added in C++11 and so was `alignas`, so we might as well use the latter, which is less error prone to use and *not* deprecated:

```suggestion
    alignas(Alignment) char storage[Size];
```

**Comment by ThePixelGamer** [comment]
That's strange that the lint check is failing as the clang-format changed it to be single lines

**Comment by ThePixelGamer** [comment]
*[`include/nn/util/util_TypedStorage.h` line 10 at r3](https://reviewable.io/reviews/open-ead/nnheaders/12#-NJ7_JiUCO6uXToWznRA-r3-10:-NJMmKCu3oLYxsi0rylD:b-896fix) ([raw file](https://github.com/open-ead/nnheaders/blob/b88fd910c3c1f96527285031422e2e7b02e514c6/include/nn/util/util_TypedStorage.h#L10)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

As previously discussed, std::aligned_storage is deprecated in modern standards so we should avoid using it (even if that's what the original nn code does). `std::aligned_storage` was added in C++11 and so was `alignas`, so we might as well use the latter, which is less error prone to use and *not* deprecated:

```suggestion
    alignas(Alignment) char storage[Size];
```
</blockquote></details>

Done.

**Comment by leoetlino** [comment]
Will merge this in 24 hours if there are no objections.

**Pull request merged**